src/AbstractApi.php

```diff
@@ -2,694 +2,395 @@
 
 namespace Apiz;
 
-use Apiz\Exceptions\UnknownResponseClassException;
 use Apiz\Http\AbstractClient;
 use Apiz\Http\Clients\GuzzleClient;
+use Apiz\Http\Request;
 use Apiz\Http\Response;
+use Apiz\Traits\Hookable;
+use Exception;
 
 /**
  * Class AbstractApi
  * @package Apiz
  */
 abstract class AbstractApi
 {
+    use Hookable;
+
     /**
      * list of available http exceptions
      *
      * @var array
      */
     protected $httpExceptions = [];
 
-
     /**
      * skip exception when its value true
      *
      * @var bool
      */
-    protected $skipHttpException = false;
-
-    /**
-     * Options for guzzle clients
-     *
-     * @var array
-     */
-    protected $options = [];
-
-    /**
-     * guzzle base URL
-     *
-     * @var string
-     */
-    protected $baseUrl = '';
-
-    /**
-     * URL prefix
-     *
-     * @var string
-     */
-    protected $prefix = '';
-
+    protected $shouldSkipHttpException = false;
 
     /**
      * this variable contains request details
      *
-     * @var array
+     * @var Request
      */
-    protected $request = [];
-
-
-    /**
-     * Default headers options for request
-     *
-     * @var array
-     */
-    protected $defaultHeaders = [];
-
-    /**
-     * Default Query options for request
-     *
-     * @var array
-     */
-    protected $defaultQueries = [];
-
-
-    /**
-     * when need to skip default header make it true
-     *
-     * @var bool
-     */
-    protected $skipDefaultHeader = false;
-
-    /**
-     * when need to skip default query make it true
-     *
-     * @var bool
-     */
-    protected $skipDefaultQueries = false;
-
-
-    protected $preHookFn = null;
-
-    /**
-     * The main client to do all the magic
-     *
-     * @var AbstractClient
-     */
-    protected $client;
-
-    /**
-     * Request parameters
-     *
-     * @var array
-     */
-    protected $parameters = [];
+    protected $request;
 
     /**
      * contains custom response
      *
-     * @var $response
-     */
-    protected $response = Response::class;
-    /**
-     * @var callable
-     */
-    protected $successHookFn = null;
-    /**
-     * @var callable
-     */
-    protected $failsHookFn = null;
-
-    public function __construct()
-    {
-        $this->baseUrl = $this->baseUrl();
-
-        $this->defaultHeaders = $this->setDefaultHeaders();
-        $this->defaultQueries = $this->setDefaultQueries();
-
-        $this->client = $this->getClient();
-    }
-
-    protected function getClient()
-    {
-        return new GuzzleClient();
-    }
-
-    /**
-     * set base URL for guzzle client
-     *
-     * @return string
+     * @var Response
      */
-    abstract protected function baseUrl();
-
+    protected $response;
 
     /**
-     * set url prefix from code
-     * @param string $prefix
-     * @return null|string
+     * AbstractApi constructor.
      */
-    protected function setPrefix($prefix = '')
+    public function __construct()
```

Comment:
It would be better and more testable if we pass Request object from constructor. 
Ex: __construct(Request $request = null);
If $request is null then you can instantiate the Request class.

Reply:
Good idea. Updated.